### ISSUE TITLE
[Feature Request] Behavior control of arrows passing through receptors

### ISSUE THREAD

**Issue by waiei** [TeamRizu/OutFox#647]
### Please Select the game mode your feature request is about.

All Game Modes

### Is your feature request related to a problem?

Currently, when a tap passes through a receptor in dance mode due to a miss decision or other reasons, it is painted black
This behavior is similar to that of DDR EXTREME and earlier, and is very difficult to see in the case of dense music

![image](https://user-images.githubusercontent.com/8708955/215028982-bf22159d-41fa-4800-99ce-f952e25b6ef5.png)

### Describe the solution you'd like

In DDR SuperNOVA and later, arrows that pass through receptors gradually become transparent
Since Project OutFox supports a variety of game modes It would be useful to be able to control this behavior on the note skin side

Since the behavior differs depending on the type of arrow, it is important to consider how much control should be available

Reference video in the case of DDR
https://www.youtube.com/watch?v=-MB3i81YUrg
- The BPM doubles at `0:21` and `0:39`, but they appear to be transparent in the same way
  - I wonder if the transparency is changing by the number of pixels instead of the time after passing through the receptor
- Long notes and shock arrows pass through without becoming transparent
  - In the case of Project OutFox, I think Hold, Roll, Mine, and Long mine are equivalent to this
- `1:54` Taps at the same position as the beginning of a long note do not become transparent, but pass through without becoming transparent
  - In Project OutFox, it may be necessary to decide whether to change the behavior by setting "DrawHoldHeadForTapsOnSameRow" or not

### Describe alternatives you've considered

I think it would be good to be able to freely control the behavior of filling in black with NoteSkin commands, even if it is not possible to control it so finely

#### Implementation I imagine
`metrics.ini`
```ini
# Count or Time (Time: seconds/Count: beats)
PassedReceptorEffectType=Count
# 0.5 beats after the arrow passes through the receptor, diffusealpha,0
PassedReceptorEffectTime=0.5
PassedReceptorEffectCommand=diffusealpha,0
```

**Comment by Scraticus** [comment]
This is a good idea, we will look into it

**Comment by JoseVarelaP** [comment]
Testing out the implementation.


Using Beat timing to fade out
```ini
PassedReceptorTransparentNote=1
PassedReceptorBeatsUntilZeroColor=0.5
PassedReceptorUseTime=0
```
https://github.com/TeamRizu/OutFox/assets/23246027/57e6ba99-748f-4415-97ba-959691e76f8e

Using Seconds timing to fade out
```ini
PassedReceptorTransparentNote=1
PassedReceptorBeatsUntilZeroColor=0.5
PassedReceptorUseTime=1
```
https://github.com/TeamRizu/OutFox/assets/23246027/441543e2-a237-4816-92d5-920dc0e86c0c

**Comment by waiei** [comment]
@JoseVarelaP 
Thank you, I think it is very good!
By the way, if I want to fill it with black as before, is it ok to use assedReceptorTransparentNote=0?
I was a little concerned that in Dance mode, either black paint or transparency is fine, but in other game modes, we may see a different pattern
In that case, =2,=3, ... and will it be extended?
Nevertheless, I think most games will be fine with the behavior you have addressed, and I look forward to seeing the implementation reflected!

**Comment by JoseVarelaP** [comment]
@waiei 
>I was a little concerned that in Dance mode, either black paint or transparency is fine, but in other game modes, we may see a different pattern

So they way I've implemented is dependent on the **NoteSkin**. So each noteskin can decide for itself if it wants to draw missed notes transparently or opaque.

> By the way, if I want to fill it with black as before, is it ok to use `PassedReceptorTransparentNote=0`?

Yes, by default it is 0, which is the "fade to black" logic.

> In that case, =2,=3, ... and will it be extended?

Not currently, it is just a boolean value.

**Comment by waiei** [comment]
Okay, I will try different things once the implementation is reflected
Thanks!

**Comment by waiei** [comment]
I confirmed that the arrow fades out and disappears as it passes through the receptor.
However, the speed at which it disappears depends on the high speed setting.

The following movie shows the operation with DDR GRANDPRIX, and it fades out at the same speed as one beat in x1.0, regardless of the high speed setting.

https://www.youtube.com/watch?v=g5c-B_pGDiA

Is it possible to make this behavior?

**Comment by waiei** [comment]
Also, the PassedReceptorBeatsUntilZeroColor setting does not appear to match the actual beats.
The following image is a comparison at x1.0, which is assumed to fade out at one beat position when PassedReceptorBeatsUntilZeroColor=1. However, in reality, the assumed behavior is not achieved unless PassedReceptorBeatsUntilZeroColor=0.125.

![Image](https://github.com/user-attachments/assets/ec932bcc-3aa2-4ac1-9d07-babbbeadcf01)

**Comment by JoseVarelaP** [comment]
Wrote up a new implementation that does true pixel calculation to fade out the notes. However, this does replace the current `PassedReceptorBeatsUntilZeroColor` to `PassedReceptorPixelsUntilZeroColor`. In this example video, it goes to 90 pixels.

https://github.com/user-attachments/assets/8ba6b4ab-fe35-49b2-8f7d-4d941609d51e

**Comment by waiei** [comment]
Thank you! I think it is very good!

I understand that the definition of metrics will change from `PassedReceptorBeatsUntilZeroColor` to `PassedReceptorPixelsUntilZeroColor`.
By the way, I think the `ScreenHeight` setting is different for each theme, but does it work the same with the same value regardless of the resolution?

**Comment by JoseVarelaP** [comment]
> Thank you! I think it is very good!
> 
> I understand that the definition of metrics will change from `PassedReceptorBeatsUntilZeroColor` to `PassedReceptorPixelsUntilZeroColor`. By the way, I think the `ScreenHeight` setting is different for each theme, but does it work the same with the same value regardless of the resolution?

It does scale, as the calculation is dependent on the NoteField's scale, which the engine/themes already take care of, so a setting of 90px scales to a theme in 480p (GrooveNights) to another one in 1080p (XX starlight)

**Comment by waiei** [comment]
I confirmed on Discord that the parameter name has been updated to `PassedReceptorValueUntilZeroColor`. ( [Discord link](https://discord.com/channels/422897054386225173/439809793964769291/1449306656952553484) )
It's working perfectly now as I expected! Thank you for the fix.

**Comment by JoseVarelaP** [comment]
Yeah, I should've noted it here when I updated it... ^^; But I'm glad you liked the fix!
